prime-router/src/main/kotlin/azure/EmailEngineFunction.kt

```diff
@@ -0,0 +1,328 @@
+            }
+        }
+    }
+
+    /**
+     * Determine if a schedule should fire
+     */
+    private fun shouldFire( schedule: EmailSchedule ): Boolean{
+
+        val parser = CronParser( CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX) );
+        // Get date for last execution
+        val now = ZonedDateTime.now();
+        val executionTime = ExecutionTime.forCron(parser.parse(schedule.cronSchedule));
+        val timeFromLastExecution = executionTime.timeFromLastExecution(now);
+
+        return ( timeFromLastExecution.get().toSeconds() <= 4*60);
+    }
+
+    /**
+     * Reports the last time that the timer has been fired for this schedule
+     * 
+     * @param schedule the schedule to check the last time fired against
+     * @returns Date of the last time fired
+     */
+    private fun getLastTimeFired( schedule: EmailSchedule ): Date {
+        val parser = CronParser( CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX) );
+        // Get date for last execution
+        val now = ZonedDateTime.now();
+        val executionTime = ExecutionTime.forCron(parser.parse(schedule.cronSchedule));
+
+        return Date.from(executionTime.lastExecution(now).get().toInstant());
+    }
+
+    /**
+     * Retreives the organization to send the emailschedule to
+     * 
+     * @param schedule the schedule to check organizations against
+     * @returns List of organizations for the schedule
+     */
+    private fun getOrgs( schedule: EmailSchedule ): List<String> {
+        return (
+            if (schedule.organizations.size > 0) 
+                schedule.organizations 
+            else 
+                fetchAllOrgs() 
+        )
+    }
+
+    /**
+     * Retrieves the list of all organization supported
+     * 
+     * @returns List of all organizations supported
+     */
+    private fun fetchAllOrgs(): List<String>{    
+        @Suppress( "NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER" )    
+        var ret = workflowEngine.db.transactReturning {  tx -> 
+            @Suppress( "UNRESOLVED_REFERENCE")
+            val settings = workflowEngine.db.fetchSettings( SettingType.ORGANIZATION, tx )
+            @Suppress( "NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER" )
+            settings.map{ it.getName() }
+        }
+       
+        @Suppress("OVERLOAD_RESOLUTION_AMBIGUITY")
+        System.out.println( ret );
+        return ret;
+    }
+
+
+    @FunctionName("createEmailSchedule")
+    @StorageAccount("AzureWebJobsStorage")
+    fun createEmailSchedule(
+        @HttpTrigger(
+            name = "createEmailSchedule",
+            methods = [HttpMethod.POST],
+            authLevel = AuthorizationLevel.ANONYMOUS,
+            route = "email-schedule"
+        ) request: HttpRequestMessage<String?>,
+        context: ExecutionContext,
+    ): HttpResponseMessage {
+        var user:String? = validateUser( request, context );
+        var ret = request.createResponseBuilder(HttpStatus.CREATED);
```

Comment:
This is a bit nit-picky but I'll raise it nonetheless: this code assumes success by default and then changes that status if things went badly.
I would encourage the opposite:
```kotlin
        var user:String? = validateUser( request, context );
        var ret = request.createResponseBuilder(HttpStatus.UNAUTHORIZED);

        if( user !== null ){
            WorkflowEngine.databaseAccess.insertEmailSchedule(request.body, user );
            ret.status(HttpStatus.CREATED )
        }

        return ret.build();
```

This eliminates the else branch and gives a better default return code.

Reply:
Agreed - done